### ISSUE TITLE
'Left Off Here' Page Flag

### ISSUE THREAD

**Issue by lmullane** [bcgov/foi-flow#3409]
* As an IAO analyst
* I want a ''Left Off Here' page flag to mark the last page I was on in a records package
* so that I easily return to that page when I come back to the records package

**Assumptions & Scope**	
What are the assumptions for this story?

Left Off here will be for the entire records package not the individual documents in the package?? Could we do a flag for each document? Would that be useful?

What is IN scope?
Marking a page as Left Off here
The page 
A button to return to the page

What is NOT in scope?
Selecting other page flags and opening the page flag drop down
Filtering by page flag

**Acceptance Criteria**	

**Scenario 1: **
* GIVEN the analyst has clicked on a record page in the left hand navigitation
* AND opened the page flag drop down
* WHEN the analyst clicks 'Left Off Here'
* THEN the page selected will be flagged with the "Left Off Here' icon

**Scenario 2: No page has been flagged with 'Left Off Here'**
* GIVEN the analyst is in a records package
* WHEN no page has been flagged as 'Left Off Here'
* THEN the 'Page Left Off Here' button in the toolbar will be inactive 

**Scenario 3: Page has been flagged with 'Left Off Here'**
* GIVEN the analyst is in a records package
* WHEN a page has been flagged as 'Left Off Here'
* THEN the 'Page Left Off Here' button in the toolbar will be active 

**Scenario 4: Click Page 'Left Off Here' button**
* GIVEN the analyst is in a records package
* AND a page has been flagged as 'Left Off Here'
* WHEN the analyst clicks the active 'Page Left Off Here' button in the toolbar 
* THEN the analyst will automatically be navigated to the page with the 'Left Off Here' flag in the editing pane
* AND the flagged page will be scrolled too in the Panel (i.e. the thumbnail of the pages)
* AND the flagged page will be selected in the Organize By left panel

**Dependencies? What is the impact of this dependency? (If so, link dependency in the ticket, make it visible in a team´s backlog)**

**Validation Rules? (If yes, list here)**	

**Design**
@xxx - please link the Design here

**Definition of Ready**

1. [ ] Is there a well articulated User Story?
2. [ ] Is there Acceptance Criteria that covers all scenarios (happy/sad paths)?
3. [ ] If there is a user interface, is there a design? 
4. [ ] Does the user story need user research/validation?
5. [ ] Does this User Story needs stakeholder approval?
6. [ ] Design / Solution accepted by Product Owner
7. [ ] Is this user story small enough to be completed in a Sprint? Should it be split?
8. [ ] Are the dependencies known/ understood? (technical, business, regulatory/policy)
9. [ ] Has the story been estimated?

**Definition of Done**

1.    [ ] Passes developer unit tests
2.    [ ] Passes peer code review
3.    [ ] If there's a user interface, passes UX assurance
4.    [ ] Passes QA of Acceptance Criteria with verification in Dev and Test
5.	[ ] Confirm Test cases built and succeeding 
6.	[ ] No regression test failures
7.	[ ] Test coverage acceptable by Product Owner
8.	[ ] Ticket ready to be merged to master or story branch
9.	[ ] Developer to list Config changes/ Update documents and designs
10.	[ ] Can be demoed in Sprint Review 
11.	[ ] Tagged as part of a Release
12.   [ ] Feature flagged if required
13.  [ ] Change Management activities done?

**Comment by lmullane** [comment]
@JHarrietha-AOT, could you please review and add the link to the designs.

@KyEggleston, please review as well.

I have gone with the assumption that the analysts is in control and must mark a page as 'Left Off Here,' rather than the system doing so automatically. If you feel it should be automatic let me know?

Cc @abin-aot

**Comment by JHarrietha-AOT** [comment]
@lmullane Did we want to add the module that Jamie suggested at the Branch Meeting where if an analyst is exiting the redaction app and hasn't applied the "Left Off Here" page flag it prompts them to apply it? Or leave that out of the MVP version on PDF Tron?

If not everything else looks good.

**Comment by lmullane** [comment]
Good catch, @JHarrietha-AOT. I've added the warning modal scenarios.

**Comment by KyEggleston** [comment]
Looks good to me, and I agree flagging should be done manually (rather than automatically). The warning @JHarrietha-AOT is a good addition

**Comment by m-prodan** [comment]
Looks good @lmullane - just one suggestion:

GIVEN the analyst as not applied a 'Left Off Here' flag
AND the analyst has made an annotation in records package (i.e., comment, redaction, etc.)
WHEN the analyst goes to exits the redaction app
THEN the a warning modal will open with the message: Do you want to close the records package without adding a 'Left Off Here' flag?
AND the modal will have a 'Close' button

Should it be continue and cancel? IE - cancel will close the modal and return to the app, continue will let them exit the redaction app w/o flagging a page as left off here. Would need another scenario to address both paths

**Comment by lmullane** [comment]
As discussed, Left Off Here will be unique and independent of Page Flags. I will revise the AC upon receiving updated designs from @JHarrietha-AOT. 

cc @m-prodan, @abin-aot

**Comment by abin-aot** [comment]
Adding the assumptions from tech refinement, 
>> if any flags applied on a page other than "left off page" , then Left off cant be applied.
>> Page left Off on the top panel of PDFTron will be removed
>> Loren/Business to rewrite the AC, Jacky to update UX

**Comment by abin-aot** [comment]
Adding the assumptions from tech refinement, 
>> if any flags applied on a page other than "left off page" , then Left off cant be applied.
>> Page left Off on the top panel of PDFTron will be removed
>> Loren/Business to rewrite the AC, Jacky to update UX

**Comment by lmullane** [comment]
@JHarrietha-AOT, did we remove the 'Left Off Here' button from the designs?

**Comment by JHarrietha-AOT** [comment]
@lmullane Yes, button removed. It is in context menu and filter will be used to navigate to it

**Comment by lmullane** [comment]
@JHarrietha-AOT, please review and advise if I have correctly captured the design update in the revised AC.

**Comment by aparna-aot** [comment]
@KyEggleston Except Scenarios 2 & 3 (related to filters), this ticket is also is in test-marshal & is Ready For QA.
All the filter related scenarios will be covered in the task - Page Flag Filter and Views#3417.

**Comment by KyEggleston** [comment]
Will wait for #3417 to test Scenario 2 and 3. 

One note for Scenario 5 and 7 - currently this warning shows up for when a user is trying to sign out of the Doc Reviewer (i.e., selecting "Sign Out"). 

However, I think the more likely process for users to exit the Doc Reviewer will be through their browser. How can we handle these situations? Currently, no warning modal pops up if they try to close their Doc Reviewer window and there is no 'Left Off Here' flag assigned. Is that something we want / can do? @lmullane @m-prodan @aparna-aot

**Comment by lmullane** [comment]
Do we have a warning if you try to close a browser window and have not saved changes in FOI Mod? For example, on the request view.

If so, we should have the same behaviour here, but would be a separate, future task (aka not part of the release).

**Comment by aparna-aot** [comment]
@lmullane @KyEggleston Yes, We have JavaScript Alerts showing up on FOI MOD if you try to close the window or click back button. 
We could implement the same alert. But showing the above modal will be tricky as JS will show its own native alert first & then displays our modal.

This is the existing alert in FOI MOD:
![image.png](https://images.zenhubusercontent.com/61a7e1d68c56cdce5dc09fd8/e9d34383-6131-408c-b861-8a82db881551)

**Comment by KyEggleston** [comment]
All ACs pass, closing ticket.
